executor/aggregate.go

```diff
@@ -386,32 +389,134 @@ func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext, f
 }
 
 // getGroupKey evaluates the group items and args of aggregate functions.
-func (w *HashAggPartialWorker) getGroupKey(sc *stmtctx.StatementContext, row chunk.Row) ([]byte, error) {
-	w.groupValDatums = w.groupValDatums[:0]
+func (w *HashAggPartialWorker) getGroupKey(sc *stmtctx.StatementContext, input *chunk.Chunk) (err error) {
+	numRows := input.NumRows()
+	for i := 0; i < numRows; i++ {
+		w.groupKey[i] = w.groupKey[i][:0]
+	}
 	for _, item := range w.groupByItems {
-		v, err := item.Eval(row)
+		tp := item.GetType()
+		buf, err := globalColumnAllocator.get(tp.EvalType(), numRows)
```

Comment:
Since types of group-by columns won't change during processing, we can allocate columns for all `HashAggPartialWorker` once when open and put them back when we close these workers.